primitives/account/src/account/staking_contract/mod.rs

```diff
@@ -177,4 +182,97 @@ impl StakingContract {
 
         slots_builder.build()
     }
+
+    /// Returns the amount of active stake (i.e., total stake of active validators).
+    pub fn get_active_stake(&self) -> Coin {
+        self.active_validators.values().copied().sum()
+    }
+
+    /// Returns the total amount of coins that are supporting the upgrade and are active.
+    /// The support is determined by a function over the signal data.
+    /// IMPORTANT: This is a fairly expensive function, iterating over all validators.
+    pub fn get_supporting_stake<
+        T: DataStoreReadOps + DataStoreIterOps,
+        F: Fn(Option<Blake2bHash>) -> bool,
+    >(
+        &self,
+        data_store: &T,
+        support_check: F,
+    ) -> Coin {
+        StakingContractStoreRead::new(data_store)
+            .iter_validators()
+            .filter_map(|validator| {
+                if validator.is_active() && support_check(validator.signal_data) {
+                    Some(validator.total_stake)
+                } else {
+                    None
+                }
+            })
+            .sum()
```

Comment:
Alternatively:
```suggestion
        let staking_Contract_read = StakingContractStoreRead::new(data_store);
        self.active_validators
            .iter()
            .filter_map(|(validator_address, stake)| {
                let validator_signal_data = staking_contract_read
                    .get_validator(address)
                    .expect("Active Validators must be present in the Staking Contract.")
                    .signal_data;
                support_check(validator_signal_data).then_some(*stake)
            })
            .sum()
```
Should also work. It shifts from going over all validator into going over only active validators and querying their signal data individually. May even be worse, just left it here for consideration.

Reply:
Depends on the number of validators. It's `log(n) + n` for the current vs. `m * log(n)` in the other case. Not sure what is better in practice.